### PR TITLE
[6.0.1] Add new Foundation libraries to SwiftRuntimeLibsOrdered and update others (#76019)

### LINKED ISSUE #76555: Significant build regressions on `swift:6.0-noble` compared to `5.10-noble`
## ~~Absolutely massive~~ Significant build regressions on linux ~~, worsened by swift-testing~~

**EDIT: Please also read my [next comment](https://github.com/swiftlang/swift/issues/76555#issuecomment-2362522356) which includes more info.**

### Environment
* GitHub Actions
* 8 CPU x 16GB RAM c7x EC2 Instance (we use [RunsOn](https://runs-on.com/))
* Migrating from `swift:5.10-noble` to `swift:6.0-noble`
* Tried with and without using a previous `.build` cache. No behavior difference noticed at all.
* 200K-300K LOC project
* Updated to `// swift-tools-version:6.0` in `Package.swift`
* All targets still have Swift 5 language mode enabled.
* No actual tests migrated to swift-testing just yet.

### What Happened?
* Our Tests CI started getting stuck after the toolchain update.
* The CI machines were getting killed by AWS. Considering this has been a common problem specially in the Server Side Swift ecosystem, which also has been brought up for ages and numerous times, my first guess was that the machine is **running out of RAM** when building the Swift project.
* After a lot of tries, I noticed a `c7x`-family machine with **64 CPU x 128 GB RAM (8x larger than before)** runs the tests as well as they were being run before, on swift 5.10.
* My first guess was that maybe Swift testing is causing an issue, so tried `--disable-swift-testing`.
* After that, our tests have been running on an only 2x larger machine, and the rest of the things are back to normal in tests CI.
* With this bigger machine, it still almost takes as long as before for the tests CI to complete (we run tests in parallel with 64 workers).
* So this means that not only Swift 6 has a significant build regression, but also swift-testing makes it go from `significant` to `absolutely massive`.
* This is borderline unusable for the Swift-Server ecosystem. I hope this problem requires more specific situations to happen that what meets the eye.
* Still trying things out with the deployment CI. even 2x CPU and 4x RAM isn't proving helpful even though I did throw the `--disable-swift-testing` flag into the mix.

### Reproduction

Not sure.

### Expected behavior

No regressions. Preferably even faster than previous Swift versions.

### Environment

Mentioned above.

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by finagolfin**
__Explanation:__ Changes a list of automatically de-duped libraries passed to the linker to reduce memory allocated while linking new Foundation and runtime libraries

__Scope:__ Updates a list of libraries that already contains Foundation to add additional Foundation and runtime libraries

__Issue:__ apple/swift-foundation#872

__Original PRs:__ #76019 and #76224

__Risk:__ Low - scope is narrow and tested with the already existing Foundation and runtime libraries in the list

__Testing:__ Testing done via swift-ci testing

__Reviewer:__ @Azoy @etcwilde @jmschonfeld

There was some concern from others about removing the libicu de-duplication in #76224 for this 6.0 branch, but I'm fairly certain that is entirely unused after the Foundation re-core to use _FoundationICU instead, as I noted there.

@shahmishal, this may fix #76555.

### CODE REVIEW

**Comment by shahmishal** [comment]
I would like to hear @parkera thoughts on pulling this in to 6.0.1 release.

**Comment by parkera** [comment]
How much does this actually address the overall problem reported in #76555?

**Comment by finagolfin** [comment]
I don't know. Without @MahdiBM trying the fixed snapshot builds I've specified there, I can't say if this is the fix.

**Comment by shahmishal** [comment]
I would prefer we get this into `release/6.0` branch before pulling it into `release/6.0.x` branches.

**Comment by finagolfin** [comment]
@shahmishal, the main commit was already merged into `release/6.0` in #76504 recently. I simply added #76224 to this 6.0.1 pull because it may also help, and I'm submitting it to `release/6.0` next anyway.

If you want, I can change this 6.0.1 pull to only include #76019, which is already in the 6.0 branch.

**Comment by finagolfin** [comment]
> added https://github.com/swiftlang/swift/pull/76224 to this 6.0.1 pull because it may also help, and I'm submitting it to release/6.0 next anyway.

Submitted in #76613.

**Comment by MahdiBM** [comment]
On it 🙂

**Comment by MahdiBM** [comment]
@finagolfin trying on these:
```
- main-snapshot-2024-09-04
- main-snapshot-2024-08-29
```
Please let me know if these are NOT what you meant.

**Comment by MahdiBM** [comment]
@finagolfin unfortunately i get:
```
error: upcoming feature 'StrictConcurrency' is already enabled as of Swift version 6
```
Because of a dependency.
Not sure how to bypass this. The package is a Swift 6 package with all targets in Swift 5 mode.

**Comment by finagolfin** [comment]
I haven't messed with the language mode settings myself. Alternately, you can try with the older July 19 and July 21 snapshot builds of the 6.0 toolchain that I mentioned first, as the former was before these new Foundation libraries was added.

**Comment by MahdiBM** [comment]
No luck.
```bash
          - 6.0-snapshot-2024-07-19
          - 6.0-snapshot-2024-07-21
```
Complains about `CompilerPluginSupport` or whatever

**Comment by bnbarham** [comment]
Are you able to share your `Package.swift`? Not clear on where `StrictConcurrency` would be being enabled here.

**Comment by MahdiBM** [comment]
It's from pointfree's Dependency package which we depend on IIRC. Not our package.

**Comment by bnbarham** [comment]
> It's from pointfree's Dependency package which we depend on IIRC. Not our package.

Which one?

**Comment by MahdiBM** [comment]
https://github.com/pointfreeco/swift-dependencies

**Comment by MahdiBM** [comment]
Perhaps my dependencies are out of date? trying with addition of `swift package update`.

**Comment by MahdiBM** [comment]
Ok yeah ... there was a recent release and our dependencies were not up to date (actually now I notice GitHub's Dependabot has stopped working on Swift 6?! need to report that).
After that there was also another problem in another dep which i manually fixed.

Now getting this:
```
.build/aarch64-unknown-linux-gnu/debug/CNIODarwin-tool.build/module.modulemap:1:8: error: redefinition of module 'CNIODarwin'
1 | module CNIODarwin {
  |        `- error: redefinition of module 'CNIODarwin'
2 |     umbrella header ".build/checkouts/swift-nio/Sources/CNIODarwin/include/CNIODarwin.h"
3 |     export *

.build/aarch64-unknown-linux-gnu/debug/CNIODarwin.build/module.modulemap:1:8: note: previously defined here
1 | module CNIODarwin {
  |        `- note: previously defined here
2 |     umbrella header ".build/checkouts/swift-nio/Sources/CNIODarwin/include/CNIODarwin.h"
3 |     export *
```

Of course none of these errors occur on the release Docker images.

**Comment by MahdiBM** [comment]
FWIW The builds fail at almost the same time:
```
          - main-snapshot-2024-09-04
          - main-snapshot-2024-08-29
```

**Comment by MahdiBM** [comment]
For the record, [this](https://github.com/swiftlang/swift/issues/76555#issuecomment-2366846529) is the result of comparing 07-19 vs 07-21 swift 6.0 nightlies, which I mentioned to @finagolfin.
TL;DR: 30% of the regression appears to have happened between those snapshots.

I was unable to try the trunk snapshots considering they would result in weird build failures.

And also there seems to be some extra noble-specific regressions as well.

**Comment by finagolfin** [comment]
6.0.1 was just tagged, so closing this. @MahdiBM, feel free to nominate these two trunk pulls for a subsequent patch release instead, if you can show they are causing a significant portion of your slowdown.
